include/interface/blas3_interface.hpp

```diff
@@ -133,20 +129,36 @@ cl::sycl::event _gemm(Executor& ex, char _TransA, char _TransB, IndexType _M,
         _ldc);                                                             \
   }
 #ifndef NAIVE_GEMM
+
+#if defined(DYNAMIC)
   if (ex.get_device_type() ==
-      Executor::Queue_Interface_Type::device_type::SYCL_INTEL_GPU) {
+      Executor::Queue_Interface_Type::device_type::SYCL_INTEL_GPU)
+#endif
+#if defined(DYNAMIC) || defined(INTEL_GPU)
+  {
     BIND_DATA_SIZE(1024, 4096, 1024) TO_TPARAMS(128, false, 4, 4, 16, 16);
     BIND_DATA_SIZE(10, 1024, 1024) TO_TPARAMS(128, false, 2, 2, 8, 8);
     BIND_DEFAULT TO_TPARAMS(128, false, 8, 8, 8, 8);
-  } else if ((ex.get_device_type() == Executor::Queue_Interface_Type::
-                                          device_type::SYCL_RCAR_CVENGINE) &&
-             (ex.get_device_type() == Executor::Queue_Interface_Type::
-                                          device_type::SYCL_RCAR_HOST_CPU)) {
-    BIND_DEFAULT TO_TPARAMS(32, false, 8, 8, 8, 8);
-  } else {
+  }
+#endif
+
+#if defined(DYNAMIC)
+  else if ((ex.get_device_type() ==
+            Executor::Queue_Interface_Type::device_type::SYCL_RCAR_CVENGINE) ||
+           (ex.get_device_type() ==
+            Executor::Queue_Interface_Type::device_type::SYCL_RCAR_HOST_CPU))
+#endif
+#if defined(DYNAMIC) || defined(RCAR)
+  {
+    BIND_DEFAULT TO_TPARAMS(16, false, 4, 4, 4, 4);
+  }
+#endif
```

Comment:
This combination of if/else is quite confusing. 
Seems to me you are trying not to compile altogether some combination of parameters, but is not clear what are those. Can you rewrite this to
```
#if defined(RCAR)
// Valid combinations
#elif DYNAMIC
// Valid combinations
...
```

In this particular case, if there is some replication, its is OK in principle.

In general, seems to me we simply should extract R-Car path to a separate backend all together if we specifically want to disable some paths of code at compile time (and template SFINAE cannot do the trick)

Reply:
I was trying to avoid the duplication of the code. So I will duplicate and rewrite it. It will be easier. The problem is not the backend. The problem is the launching an optimised kernel by calculating  certain compile time parameters for each backend based on run-time information. The specialisation constant will solve the issue :)

Reply:
But in general, I think it is better to have a separate backend for R-Car, due to architectural difference.